File: dataStructure/queue.cpp
#include <cstdio>
#include <iostream>

using namespace std;

template <typename T>
class Queue {
 public:
  Queue(unsigned int len) : length(len), rear(-1) {
    items = new T[length];
  }
  ~Queue() {
    delete [] items;
  }

  void push(T item) {
    if (isFull()) cerr << "Queue is full!\n";
    else {
      rear++;
      items[rear] = item;
    }
  }

  void pop() {
    if (isEmpty()) cerr << "Queue is empty!\n";
    else {
      // shift the whole array.
      for (int i = 0; i < rear; i++) items[i] = items[i + 1];
      rear--;
    }
  }

  int size() {
    return rear + 1;
  }

  T front() {
    if (isEmpty()) cerr << "Queue is empty!\n";
    return items[0];
  }

  T back() {
    if (isEmpty()) cerr << "Queue is empty!\n";
    return items[rear];
  }

  bool isEmpty() {
    return rear == -1;
  }

  bool isFull() {
    return rear == length - 1;
  }

  void display() {
    printf("[");
    for (int i = 0; i <= rear; i++) cout << ' ' << items[i];
    printf(" ]\n");
  }


 private:
  int rear;
  unsigned int length;
  T *items;
};

int main() {
  Queue arr = Queue<int>(5);

  arr.push(1);
  arr.push(2);
  arr.push(3);
  arr.display();
  arr.pop();
  arr.push(4);
  arr.display();
  printf("%d\n", arr.front());

  return 0;
}

File: dataStructure/Queue.java
import java.io.PrintStream;

class Queue<T> {
  private int length;
  private int rear;
  private T[] items;

  // `static final` defines a constant value
  static final PrintStream stdout = System.out;
  static final PrintStream stderr = System.err;

  @SuppressWarnings("unchecked")
  public Queue(int length) {
    this.length = length;
    this.rear = -1;
    this.items = (T[]) new Object[this.length];
  }

  public void enqueue(T item) {
    if (this.isFull()) stderr.println("Queue is full!");
    else {
      this.rear++;
      this.items[this.rear] = item;
    }
  }

  public void dequeue() {
    if (this.isEmpty()) stderr.println("Queue is empty!");
    else {
      // shift the whole array.
      for (int i = 0; i < this.rear; i++) this.items[i] = this.items[i + 1];
      this.rear--;
    }
  }

  public int size() {
    return this.rear + 1;
  }

  public T front() throws Error {
    if (this.isEmpty()) throw new Error("Queue is empty!");
    return this.items[0];
  }

  public T back() throws Error {
    if (this.isEmpty()) throw new Error("Queue is empty!");
    return this.items[this.rear];
  }

  public boolean isEmpty() {
    return this.rear == -1;
  }

  public boolean isFull() {
    return this.rear == this.length - 1;
  }

  public void display() {
    stdout.printf("[");
    for (int i = 0; i <= this.rear; i++) stdout.print(" " + this.items[i]);
    stdout.println(" ]");
  }
}

class SimpleQueue<T> {
  private int frontIdx;
  private int rearIdx;
  private int length;
  private T[] items;

  // `static final` defines a constant value
  static final PrintStream stdout = System.out;
  static final PrintStream stderr = System.err;

  @SuppressWarnings("unchecked")
  public SimpleQueue(int len) {
    this.frontIdx = -1;
    this.rearIdx = -1;
    this.length = len;
    this.items = (T[]) new Object[len];
  }

  public void enqueue(T item) {
    if (this.isFull()) stderr.println("Queue is full!");
    else {
      if (this.frontIdx == -1) this.frontIdx = 0;
      this.rearIdx++;
      this.items[this.rearIdx] = item;
    }
  }

  public void dequeue() {
    if (this.isEmpty()) stderr.println("Queue is empty!");
    else {
      this.frontIdx++;
      if (this.frontIdx > this.rearIdx) {
        this.frontIdx = -1;
        this.rearIdx = -1;
      }
    }
  }

  public int size() {
    if (this.isEmpty()) return 0;
    return this.rearIdx - this.frontIdx + 1;
  }

  public T front() throws Error {
    if (this.isEmpty()) throw new Error("Queue is empty!");
    return this.items[this.frontIdx];
  }

  public T back() throws Error {
    if (this.isEmpty()) throw new Error("Queue is empty!");
    return this.items[this.rearIdx];
  }

  public boolean isEmpty() {
    return this.frontIdx == -1;
  }

  public boolean isFull() {
    return this.rearIdx == this.length - 1;
  }

  public void display() {
    stdout.print("[");
    for (int i = this.frontIdx; i <= this.rearIdx; i++) stdout.print(" " + this.items[i]);
    stdout.println(" ]");
  }
}
